src/lib/misc/roughtime/roughtime.cpp

```diff
@@ -0,0 +1,370 @@
+  PK_Verifier verifier(pk, "Pure");
+  verifier.update((const uint8_t*)context, sizeof(context)); //add context including \0
+  verifier.update(m_cert_dele.data(), m_cert_dele.size());
+  return verifier.check_signature(std::vector<uint8_t>(m_cert_sig.data(), m_cert_sig.data() + m_cert_sig.size()));
+  }
+
+std::array<uint8_t, 64> nonce_from_blind(const std::vector<uint8_t>& previous_response,
+                                         const std::array<uint8_t, 64>& blind)
+  {
+  std::array<uint8_t, 64> ret;
+
+  std::unique_ptr<Botan::HashFunction> hash(Botan::HashFunction::create("SHA-512"));
+  hash->update(previous_response);
+  hash->update(hash->final());
+  hash->update(blind.data(), blind.size());
+  hash->final(ret.data());
+
+  return ret;
+  }
+
+Chain::Chain(DataSource& ds)
+  {
+  const auto chain = [&ds]()
+    {
+    std::vector<uint8_t> v;
+    size_t read = 0;
+    unsigned step = 1024;
+    for(;;)
+    {
+      v.resize(read+step);
+      auto n = ds.read(v.data() + read, step);
+      read += n;
+      if(n!=step) break;
+    }
+    v.resize(read);
+    return nlohmann::json::parse(v.begin(), v.end());
+    }();
+  for(const auto& l : chain.at("links"))
+     {
+     const auto publicKeyType = l.at("publicKeyType").get<std::string>();
+     const auto serverPublicKey = Botan::Ed25519_PublicKey(Botan::base64_decode(l.at("serverPublicKey").get<std::string>()));
+     const auto nonceOrBlind = vector_to_array<64>(Botan::unlock(Botan::base64_decode(l.at("nonceOrBlind").get<std::string>())));
+     const auto response = Botan::unlock(Botan::base64_decode(l.at("reply").get<std::string>()));
+     if(publicKeyType != "ed25519")
+        throw Not_Implemented("only ed25519 publicKeyType is implemented");
+     m_links.push_back({response, serverPublicKey, nonceOrBlind});
+     }
+  }
+std::vector<Response> Chain::responses() const
+  {
+  std::vector<Response> responses;
+  for(unsigned i = 0; i < m_links.size(); ++i)
+    {
+    const auto& l = m_links[i];
+    const auto nonce = i ? nonce_from_blind(m_links[i-1].response, l.nonce_or_blind) : l.nonce_or_blind;
+    const auto response = Response::from_bits(l.response, nonce);
+    if(!response.validate(l.public_key))
+      throw Invalid_Argument("invalid signature or public key");
+    responses.push_back(response);
+    }
+  return responses;
+  }
+void Chain::next_blind_and_nonce(RandomNumberGenerator& rng, std::array<uint8_t, 64>& nonce, std::array<uint8_t, 64>& blind) const
+  {
+    rng.randomize(blind.data(), blind.size());
+    if(!m_links.empty())
+      {
+      const auto& previous = m_links.back();
+      nonce = nonce_from_blind(previous.response, blind);
+      }
+    else
+      {
+      nonce = blind;
+      }
+  }
+void Chain::append(const std::vector<uint8_t>& response,
+                  const Ed25519_PublicKey& public_key,
+                  const std::array<uint8_t, 64>& nonce_or_blind,
+                  size_t max_chain_size)
+  {
+  if(max_chain_size <= 0)
```

Comment:
Since the max_chain_size is unsigned this should be just == 0 right?

Reply:
Yes. The check currently is just doing that. But I believe this is the right check to me made, since it still is correct if the variable is changed to signed.

Reply:
Fair enough